### PR TITLE
Use SSdirection for queen locator

### PR DESCRIPTION
**Pull request by psykzz**
## About The Pull Request

Refactors the queen locator proc to use SSdirection instead

## Changelog
:cl:
code: Xeno queen tracking is now handled by a subsystem.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/living.dm [hunk 1]
```diff
@@ -622,9 +622,8 @@ below 100 is not dizzy
 		message_admins("[key_name_admin(M)] has taken [ADMIN_TPMONTY(src)].")
 
 	M.mind.transfer_to(src, TRUE)
-	fully_replace_character_name(M.real_name, real_name)
+	M.fully_replace_character_name(M.real_name, real_name)
```
Comment (DominikPanic) [inline code/modules/mob/living/living.dm]: Merge garbage

**Comment by psykzz** [comment]
image

image

**Comment by psykzz** [comment]
@spookydonut i think this should be more explicit than that which is why i'd prefer to go this way. 
If we plan to do the same with squads, maybe it would make sense then, but until that change happens I'd prefer to continue like this.

File: code/controllers/stonedmc/subsystem/direction.dm [hunk 2]
```diff
@@ -37,26 +39,40 @@ SUBSYSTEM_DEF(direction)
...
-				H = currentrun[L][currentrun[L].len]
-				currentrun[L].len--
-				H.clear_leader_tracking()
-				if(MC_TICK_CHECK)
-					return	
+	for(var/squad_id in currentrun)
+		var/mob/living/L
+		var/mob/living/SL = leader_mapping[squad_id]
+
+		if(QDELETED(SL) || !isliving(SL))
```
Comment (Rohesie) [inline code/controllers/stonedmc/subsystem/direction.dm]: The `SL` should always be `living` and should handle in its deletion being removed from the list.
These checks shouldn't be necessary.

File: code/controllers/stonedmc/subsystem/direction.dm [hunk 3]
```diff
@@ -37,26 +39,40 @@ SUBSYSTEM_DEF(direction)
...
+			if (!clear_run(squad_id))
+				return
+
+		while(currentrun[squad_id].len)
+			L = currentrun[squad_id][currentrun[squad_id].len]
+			currentrun[squad_id].len--
+			if(QDELETED(L))
```
Comment (Rohesie) [inline code/controllers/stonedmc/subsystem/direction.dm]: Same here. If this is happening, somewhere else did an oof. Shouldn't be checking each time.

File: code/controllers/stonedmc/subsystem/direction.dm [hunk 4]
```diff
@@ -37,26 +39,40 @@ SUBSYSTEM_DEF(direction)
...
+				return
+
+/datum/controller/subsystem/direction/proc/clear_run(squad_id)
+	var/mob/living/L
+	var/mob/living/carbon/Xenomorph/X
+	var/mob/living/carbon/human/H
+	while(currentrun[squad_id].len)
+		L = currentrun[squad_id][currentrun[squad_id].len]
+		currentrun[squad_id].len--
+		if (isxeno(L))
```
Comment (Rohesie) [inline code/controllers/stonedmc/subsystem/direction.dm]: Themove the `istype` checks.
Just run `L.clear_leader_tracking()`

File: code/modules/mob/living/carbon/xenomorph/xenomorph.dm [hunk 5]
```diff
@@ -259,4 +259,31 @@
 	return
 
 /mob/living/carbon/Xenomorph/reagent_check(datum/reagent/R) //For the time being they can't metabolize chemicals.
-	return TRUE
\ No newline at end of file
+	return TRUE
+
+/mob/living/carbon/Xenomorph/update_leader_tracking(mob/living/carbon/Xenomorph/X)
+	if(!hud_used || !hud_used.locate_leader)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/xenomorph.dm]:
```suggestion
	if(!hud_used?.locate_leader)
```

File: code/modules/mob/living/carbon/xenomorph/xenomorph.dm [hunk 6]
```diff
@@ -259,4 +259,31 @@
...
+		return
+
+	var/obj/screen/LL_dir = hud_used.locate_leader
+	if(!X || xeno_caste.caste_flags & CASTE_IS_INTELLIGENT)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/xenomorph.dm]: Is there reasonable expectation of this to be called without an `X`?

File: code/modules/mob/living/carbon/xenomorph/xenomorph.dm [hunk 7]
```diff
@@ -259,4 +259,31 @@
...
+		var/area/A = get_area(src.loc)
+		var/area/QA = get_area(X.loc)
+		if(A.fake_zlevel == QA.fake_zlevel)
+			LL_dir.icon_state = "trackon"
+			LL_dir.setDir(get_dir(src, X))
+		else
+			LL_dir.icon_state = "trackondirect"
+
+/mob/living/carbon/Xenomorph/clear_leader_tracking()
+	if(!hud_used || !hud_used.locate_leader)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/xenomorph.dm]:
```suggestion
	if(!hud_used?.locate_leader)
```

File: code/modules/mob/living/carbon/xenomorph/xenomorph.dm (hunk 6 above)
Comment (psykzz) [reply code/modules/mob/living/carbon/xenomorph/xenomorph.dm]: Should never be called outside the subsystem really, so no. Will remove.

**Comment by DominikPanic** [comment]
Abhorrent amount of runtimes.
```
runtime error: Cannot read null.z
proc name: update leader tracking (/mob/living/carbon/update_leader_tracking)
  source file: carbon.dm,418
  usr: null
  src: Amelia \'Nya\' Perez (/mob/living/carbon/human)
  src.loc: the floor (167,42,3) (/turf/open/floor/almayer)
  call stack:
Amelia \'Nya\' Perez (/mob/living/carbon/human): update leader tracking(null)
Direction (/datum/controller/subsystem/direction): fire(0)
Direction (/datum/controller/subsystem/direction): ignite(0)
Master (/datum/controller/master): RunQueue()
Master (/datum/controller/master): Loop()
Master (/datum/controller/master): StartProcessing(0)
```

**Comment by Rohesie** [comment]
Conflicts.

**Comment by Rohesie** [comment]
```
[05:35:06] Runtime in direction.dm, line 51: Cannot execute null.update leader tracking().
proc name: fire (/datum/controller/subsystem/direction/fire)
src: Direction (/datum/controller/subsystem/direction)
call stack:
Direction (/datum/controller/subsystem/direction): fire(0)
Direction (/datum/controller/subsystem/direction): ignite(0)
Master (/datum/controller/master): RunQueue()
Master (/datum/controller/master): Loop()
Master (/datum/controller/master): StartProcessing(0)
```

File: code/controllers/stonedmc/subsystem/direction.dm [hunk 8]
```diff
@@ -37,54 +39,62 @@ SUBSYSTEM_DEF(direction)
...
+			L.update_leader_tracking(SL)
+			if(MC_TICK_CHECK)
+				return
+
+/datum/controller/subsystem/direction/proc/clear_run(squad_id)
+	var/mob/living/L
+	while(currentrun[squad_id].len)
+		L = currentrun[squad_id][currentrun[squad_id].len]
+		currentrun[squad_id].len--
+		if(QDELETED(L))
```
Comment (Rohesie) [inline code/controllers/stonedmc/subsystem/direction.dm]: Make sure on `living/Destroy()` things are removed from the subsystem if they are in it.

**Pull request merged**